File: src/tsqr_bench.cpp
// Copyright (c) 2020 German Aerospace Center (DLR), Institute for Software Technology, Germany
// SPDX-FileContributor: Melven Roehrig-Zoellner <Melven.Roehrig-Zoellner@DLR.de>
//
// SPDX-License-Identifier: BSD-3-Clause

#include "pitts_mkl.hpp"
#include "pitts_parallel.hpp"
#include "pitts_performance.hpp"
#include "pitts_common.hpp"
#include "pitts_multivector.hpp"
#include "pitts_multivector_random.hpp"
#include "pitts_multivector_tsqr.hpp"
#include "pitts_multivector_eigen_adaptor.hpp"
#include "pitts_tensor2.hpp"
#include "pitts_tensor2_eigen_adaptor.hpp"
#include "pitts_eigen.hpp"
#include <charconv>
#include <iostream>
#include <stdexcept>


#ifdef TYPE
using Type = TYPE;
#else
using Type = double;
#endif


int main(int argc, char* argv[])
{
  PITTS::initialize(&argc, &argv);

  using mat = Eigen::Matrix<Type, Eigen::Dynamic, Eigen::Dynamic>;
  using Chunk = PITTS::Chunk<Type>;

  if( argc != 5 && argc != 6 )
    throw std::invalid_argument("Requires 4 arguments (n m reductionFactor nIter [colBlockingSize] )!");

  long long n = 0, m = 0;
  int reductionFactor = 20, nIter = 0;
  int colBlockingSize = 0;
  std::from_chars(argv[1], argv[2], n);
  std::from_chars(argv[2], argv[3], m);
  std::from_chars(argv[3], argv[4], reductionFactor);
  std::from_chars(argv[4], argv[5], nIter);
  if( argc == 6 )
    std::from_chars(argv[5], argv[6], colBlockingSize);

  const auto& [iProc,nProcs] = PITTS::internal::parallel::mpiProcInfo();
  {
    const auto& [nFirst,nLast] = PITTS::internal::parallel::distribute(n, {iProc,nProcs});
    n = nLast - nFirst + 1;
  }

  PITTS::MultiVector<Type> M(n, m);
  randomize(M);

  PITTS::Tensor2<Type> R(m,m);
  block_TSQR(M, R, reductionFactor, true, colBlockingSize);

  PITTS::performance::clearStatistics();

double wtime = omp_get_wtime();
  for(int iter = 0; iter < nIter; iter++)
  {
    block_TSQR(M, R, reductionFactor, true, colBlockingSize);
  }
wtime = omp_get_wtime() - wtime;
  if( iProc == 0 )
    std::cout << "wtime: " << wtime << "\n";



  // compare to gramian calculation
  PITTS::Tensor2<Type> MtM(m,m);
  EigenMap(MtM).noalias() = ConstEigenMap(M).transpose() * ConstEigenMap(M);

wtime = omp_get_wtime();
  for(int iter = 0; iter < nIter; iter++)
  {
    EigenMap(MtM).noalias() = ConstEigenMap(M).transpose() * ConstEigenMap(M);
  }
wtime = omp_get_wtime() - wtime;
  if( iProc == 0 )
    std::cout << "Gramian M^TM wtime (not SYRK!): " << wtime << "\n";

  if( iProc == 0 )
  {
    Eigen::JacobiSVD<mat> svd(ConstEigenMap(R));
    //std::cout << "Result:\n" << M << "\n";
    std::cout << "singular values (new):\n" << svd.singularValues().transpose() << "\n";

    Eigen::SelfAdjointEigenSolver<mat> eig(ConstEigenMap(MtM));
    //std::cout << "Gramian eigenvalues: " << eig.eigenvalues().array().reverse().sqrt().transpose() << "\n";
    std::cout << "Difference to sqrt(Gramian eigenvalues): " << (svd.singularValues().array() - eig.eigenvalues().array().reverse().sqrt()).transpose() << "\n";
  }

  PITTS::finalize();

  return 0;
}

File: examples/tsqr_bench.py
#!/usr/bin/env python3

# Copyright (c) 2020 German Aerospace Center (DLR), Institute for Software Technology, Germany
# SPDX-FileContributor: Melven Roehrig-Zoellner <Melven.Roehrig-Zoellner@DLR.de>
#
# SPDX-License-Identifier: BSD-3-Clause

""" Compare TSQR runtime with numpy qr runtime """

__authors__ = ['Melven Roehrig-Zoellner <Melven.Roehrig-Zoellner@DLR.de>']
__date__ = '2020-10-03'

import numpy as np
import pitts_py
import functools
import timeit


def timer(func):
    """measure runtime of the decorated function"""
    @functools.wraps(func)
    def wrapper_fun(*args, **kwargs):
        wtime = timeit.default_timer()
        value = func(*args, **kwargs)
        wtime = timeit.default_timer() - wtime
        print(func.__name__, "wtime:", wtime)
        return value
    return wrapper_fun

@timer
def numpy_qr(X):
    R = np.linalg.qr(X, mode='r')

@timer
def pitts_qr(X):
    R = pitts_py.block_TSQR(X, 20, False)

if __name__ == '__main__':

    pitts_py.initialize(True)

    n = 10000000
    m = 50

    X = pitts_py.MultiVector_double(n, m)
    pitts_py.randomize(X)

    for j in range(1, m+1):
        print("n: ", n, ", m: ", j)
        X.resize(n,j)
        pitts_qr(X)
        numpy_qr(X)

    n = n//10
    m = 100
    print("n: ", n, ", m: ", m)
    X.resize(n, m)
    pitts_py.randomize(X)
    pitts_qr(X)
    numpy_qr(X)

    n = n//10
    m = 200
    print("n: ", n, ", m: ", m)
    X.resize(n, m)
    pitts_py.randomize(X)
    pitts_qr(X)
    numpy_qr(X)

    pitts_py.finalize(True)
